poweriq-worker/src/main/java/com/vmware/flowgate/poweriqworker/jobs/PowerIQService.java

```diff
@@ -564,16 +559,127 @@ public void saveSensorAssetsToFlowgate(Map<String, Asset> exsitingSensorAssets,
+         }
          offset += limit;
       }
+   }
+
+   public List<Asset> updatePduMetricformular(List<Asset> sensorAssets, Map<String,Asset> pduIdAndAssetMap){
+      List<Asset> pduAssets = new ArrayList<Asset>();
+      for(Asset sensorAsset : sensorAssets) {
+         com.vmware.flowgate.common.model.Parent parent = sensorAsset.getParent();
+         if(parent == null) {
+            continue;
+         }
+         String pduId = parent.getParentId();
+         Asset pduAsset = pduIdAndAssetMap.get(pduId);
+         int position = sensorAsset.getCabinetUnitPosition();
+         String positionInfo = FlowgateConstant.DEFAULT_CABINET_UNIT_POSITION;
+         Map<String,String> sensorAssetJustfication = sensorAsset.getJustificationfields();
+         if(position != 0) {
+            positionInfo = String.valueOf(position);
+         }else {
+            if(sensorAssetJustfication != null) {
+               String sensorInfo = sensorAssetJustfication.get(FlowgateConstant.SENSOR);
+               if(sensorInfo != null) {
+                  try {
+                     Map<String,String> sensorInfoMap = getInfoMap(sensorInfo);
+                     String positionFromPowerIQ = sensorInfoMap.get(FlowgateConstant.POSITION_FROM_POWERIQ);
+                     if(positionFromPowerIQ != null) {
+                        positionInfo = positionFromPowerIQ;
+                     }
+                  } catch (IOException e) {
+                     positionInfo = FlowgateConstant.DEFAULT_CABINET_UNIT_POSITION;
+                  }
+               }
+            }
+         }
+         Map<String, Map<String, Map<String, String>>> formulars = pduAsset.getMetricsformulars();
+         if(formulars == null || formulars.isEmpty()) {
+            formulars = new HashMap<String, Map<String, Map<String, String>>>();
+            Map<String, Map<String, String>> sensorFormulars = generateNewMetricformular(sensorAsset);
+            if(sensorFormulars.isEmpty()) {
+               continue;
+            }
+            formulars.put(FlowgateConstant.SENSOR, sensorFormulars);
+         }else {
+            Map<String, Map<String, String>> sensorFormulars = formulars.get(FlowgateConstant.SENSOR);
+            if(sensorFormulars == null || sensorFormulars.isEmpty()) {
+               sensorFormulars = generateNewMetricformular(sensorAsset);
+               formulars.put(FlowgateConstant.SENSOR, sensorFormulars);
+            }else {
+               Map<String,String> locationAndMetricMap = null;
+               switch (sensorAsset.getSubCategory()) {
+               case Humidity:
+                  locationAndMetricMap = sensorFormulars.get(MetricName.PDU_HUMIDITY);
+                  if(locationAndMetricMap == null) {
+                     locationAndMetricMap = new HashMap<String,String>();
+                  }
+                  /**
+                   * If there are many sensor assets that them cabinet unit position is 0,
```

Comment:
The comment is not right . since the position is a combine of rackunit and string based position like inlet, outlet, external, etc. not only 0.

Reply:
Removed